C error when unwrapping option arrays

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 90c638e, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 725e25991d7445f5609958816abc3370a5c0dac3.90c638e
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz  
|Memory              |3.57GB/15.86GB
|                    |
|V executable        |***
|V last modified time|2025-03-18 16:32:21
|                    |
|V home dir          |OK, value: ***
|VMODULES            |OK, value: ***
|VTMP                |OK, value: ***
|Current working dir |OK, value: ***
|                    |
|Git version         |git version 2.38.1.windows.1
|V git status        |weekly.2024.24-1728-g90c638e6
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |N/A
|clang version       |N/A
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  test.v && F:\***\test.exe`
```v
fn main() {
	mut arr := ?[]int(none);
	if arr == none {
		arr = []int{};
	} else {
		unwrap(arr) << 1;
	}
}

@[inline]
pub fn unwrap[T](t ?T) T {
	return t or { panic('unexpected `none`') }
}
```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: C:/Users/***/AppData/Local/Temp/v_0/test.01JPN6Q6G3A7QRMWMR8EF3VRJP.tmp.c:3859: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/***/AppData/Local/Temp/v_0/test.01JPN6Q6G3A7QRMWMR8EF3VRJP.tmp.c:5901: error: cannot convert 'struct array' to 'struct _option_Array_int'
...
cc: C:/Users/***/AppData/Local/Temp/v_0/test.01JPN6Q6G3A7QRMWMR8EF3VRJP.tmp.c:5901: error: cannot convert 'struct array' to 'struct _option_Array_int'
(note: the original output was 3 lines long; it was truncated to its first 2 lines + the last line)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

maybe works fine, or maybe an error message, idk



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2604,7 +2604,7 @@ fn (mut g Gen) keep_alive_call_postgen(node ast.CallExpr, tmp_cnt_save int) {
 
 @[inline]
 fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang ast.Language, is_smartcast bool) {
-	arg_typ := if arg.ct_expr {
+	mut arg_typ := if arg.ct_expr {
 		g.unwrap_generic(g.type_resolver.get_type(arg.expr))
 	} else {
 		g.unwrap_generic(arg.typ)
@@ -2744,6 +2744,13 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 		if (arg_sym.info is ast.Alias || exp_sym.info is ast.Alias) && expected_type != arg_typ {
 			g.expr_opt_with_alias(arg.expr, arg_typ, expected_type)
 		} else {
+			if arg.expr is ast.Ident {
+				if arg.expr.obj is ast.Var {
+					if arg.expr.obj.smartcasts.len > 0 {
+						arg_typ = arg.expr.obj.smartcasts.last()
+					}
+				}
+			}
 			g.expr_with_opt(arg.expr, arg_typ, expected_type)
 		}
 		return

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -1030,13 +1030,22 @@ fn (mut g Gen) infix_expr_left_shift_op(node ast.InfixExpr) {
 			elem_sym := g.table.final_sym(array_info.elem_type)
 			elem_is_array_var := elem_sym.kind in [.array, .array_fixed] && node.right is ast.Ident
 			g.write('array_push${noscan}((array*)')
+			mut needs_addr := false
 			if !left.typ.is_ptr()
 				|| (node.left_type.has_flag(.shared_f) && !node.left_type.deref().is_ptr()) {
-				g.write('&')
+				if node.left is ast.CallExpr {
+					g.write('ADDR(${g.styp(node.left_type)}, ')
+					needs_addr = true
+				} else {
+					g.write('&')
+				}
 			}
 			g.expr(node.left)
 			if node.left_type.has_flag(.shared_f) {
 				g.write('->val')
+			}
+			if needs_addr {
+				g.write(')')
 			}
 			if elem_sym.kind == .function {
 				g.write(', _MOV((voidptr[]){ ')

--- a/vlib/v/tests/options/option_unwrapped_arg_generic_opt_test.v
+++ b/vlib/v/tests/options/option_unwrapped_arg_generic_opt_test.v
@@ -3,7 +3,8 @@ fn test_main() {
 	if arr == none {
 		arr = []int{}
 	} else {
-		unwrap(arr)
+		ret := unwrap(arr)
+		assert ret == arr
 	}
 	assert arr?.len == 0
 }

--- a/vlib/v/tests/options/option_unwrapped_arg_generic_opt_test.v
+++ b/vlib/v/tests/options/option_unwrapped_arg_generic_opt_test.v
@@ -1,0 +1,16 @@
+fn test_main() {
+	mut arr := ?[]int(none)
+	if arr == none {
+		arr = []int{}
+	} else {
+		unwrap(arr) << 1
+	}
+	assert arr?.len == 0
+}
+
+fn t(a ?int) {}
+
+@[inline]
+pub fn unwrap[T](t ?T) T {
+	return t or { panic('unexpected `none`') }
+}

--- a/vlib/v/tests/options/option_unwrapped_arg_generic_opt_test.v
+++ b/vlib/v/tests/options/option_unwrapped_arg_generic_opt_test.v
@@ -3,7 +3,7 @@ fn test_main() {
 	if arr == none {
 		arr = []int{}
 	} else {
-		unwrap(arr) << 1
+		unwrap(arr)
 	}
 	assert arr?.len == 0
 }

